File: src/cli/discovery.zig
const std = @import("std");

/// Supported config file formats.
pub const ConfigFormat = enum {
    json,
    toml,
};

/// Config filenames checked in priority order (per locked decision).
const config_filenames = [_][]const u8{
    ".complexityguard.json",
    "complexityguard.config.json",
    ".complexityguard.toml",
    "complexityguard.config.toml",
};

/// User-level config paths checked in priority order.
const user_config_paths = [_][]const u8{
    "complexityguard/config.json",
    "complexityguard/config.toml",
};

/// Maximum directory depth to prevent infinite loops.
const max_search_depth = 100;

/// Discovers config file path using upward search with .git boundary,
/// XDG fallback, or explicit override.
///
/// Search order:
/// 1. If explicit_path is provided, return it (after verifying existence)
/// 2. Search upward from CWD for config files, stop at .git boundary
/// 3. Fall back to user config in XDG config directory
/// 4. Return null if no config found
///
/// Returns owned string that caller must free.
pub fn discoverConfigPath(allocator: std.mem.Allocator, explicit_path: ?[]const u8) !?[]const u8 {
    // If explicit path provided, verify it exists
    if (explicit_path) |path| {
        std.fs.cwd().access(path, .{}) catch {
            return error.ConfigFileNotFound;
        };
        return try allocator.dupe(u8, path);
    }

    // Try project config (upward search)
    if (try searchProjectConfig(allocator)) |path| {
        return path;
    }

    // Try user config (XDG fallback)
    if (try searchUserConfig(allocator)) |path| {
        return path;
    }

    // No config found
    return null;
}

/// Searches upward from CWD for config files, stopping at .git boundary.
fn searchProjectConfig(allocator: std.mem.Allocator) !?[]const u8 {
    var current_dir = try std.process.getCwdAlloc(allocator);
    defer allocator.free(current_dir);

    var iterations: usize = 0;
    while (iterations < max_search_depth) : (iterations += 1) {
        // Check for config files in priority order
        for (config_filenames) |filename| {
            const config_path = try std.fs.path.join(allocator, &.{ current_dir, filename });
            defer allocator.free(config_path);

            std.fs.cwd().access(config_path, .{}) catch {
                continue; // File doesn't exist, try next
            };

            // Found config file
            return try allocator.dupe(u8, config_path);
        }

        // Check if we hit .git boundary
        const git_path = try std.fs.path.join(allocator, &.{ current_dir, ".git" });
        defer allocator.free(git_path);

        var is_git_boundary = false;
        std.fs.cwd().access(git_path, .{}) catch {
            // .git doesn't exist, continue search
        };
        // Check if .git exists
        const git_stat = std.fs.cwd().statFile(git_path) catch null;
        if (git_stat != null) {
            is_git_boundary = true;
        }

        if (is_git_boundary) {
            // Hit git boundary, stop searching
            return null;
        }

        // Move to parent directory
        const parent = std.fs.path.dirname(current_dir) orelse break;
        const parent_copy = try allocator.dupe(u8, parent);
        allocator.free(current_dir);
        current_dir = parent_copy;
    }

    return null;
}

/// Searches for user-level config in XDG config directory.
fn searchUserConfig(allocator: std.mem.Allocator) !?[]const u8 {
    const config_home = try getConfigHome(allocator) orelse {
        return null; // No XDG config directory
    };
    defer allocator.free(config_home);

    for (user_config_paths) |relative_path| {
        const full_path = try std.fs.path.join(allocator, &.{ config_home, relative_path });
        defer allocator.free(full_path);

        std.fs.cwd().access(full_path, .{}) catch {
            continue; // File doesn't exist, try next
        };

        // Found user config
        return try allocator.dupe(u8, full_path);
    }

    return null;
}

/// Gets XDG config home directory.
/// Returns $XDG_CONFIG_HOME if set, otherwise ~/.config on Unix or %APPDATA% on Windows.
fn getConfigHome(allocator: std.mem.Allocator) !?[]const u8 {
    // Check XDG_CONFIG_HOME environment variable
    const xdg_config = std.process.getEnvVarOwned(allocator, "XDG_CONFIG_HOME") catch |err| switch (err) {
        error.EnvironmentVariableNotFound => null,
        else => |e| return e,
    };
    if (xdg_config) |val| return val;

    // Fall back to HOME/.config on Unix-like systems
    const home = std.process.getEnvVarOwned(allocator, "HOME") catch |err| switch (err) {
        error.EnvironmentVariableNotFound => null,
        else => |e| return e,
    };
    if (home) |home_val| {
        defer allocator.free(home_val);
        return try std.fs.path.join(allocator, &.{ home_val, ".config" });
    }

    // Fall back to APPDATA on Windows
    const appdata = std.process.getEnvVarOwned(allocator, "APPDATA") catch |err| switch (err) {
        error.EnvironmentVariableNotFound => null,
        else => |e| return e,
    };
    if (appdata) |val| return val;

    return null;
}

/// Detects config format based on file extension.
pub fn detectConfigFormat(path: []const u8) ConfigFormat {
    if (std.mem.endsWith(u8, path, ".json")) {
        return .json;
    } else if (std.mem.endsWith(u8, path, ".toml")) {
        return .toml;
    }
    // Default to JSON if extension is unclear
    return .json;
}

// TESTS

test "detectConfigFormat returns .json for .json files" {
    try std.testing.expectEqual(ConfigFormat.json, detectConfigFormat(".complexityguard.json"));
    try std.testing.expectEqual(ConfigFormat.json, detectConfigFormat("complexityguard.config.json"));
    try std.testing.expectEqual(ConfigFormat.json, detectConfigFormat("/path/to/config.json"));
}

test "detectConfigFormat returns .toml for .toml files" {
    try std.testing.expectEqual(ConfigFormat.toml, detectConfigFormat(".complexityguard.toml"));
    try std.testing.expectEqual(ConfigFormat.toml, detectConfigFormat("complexityguard.config.toml"));
    try std.testing.expectEqual(ConfigFormat.toml, detectConfigFormat("/path/to/config.toml"));
}

test "discoverConfigPath with explicit path returns that path" {
    const allocator = std.testing.allocator;

    // Create temp file
    var tmp_dir = std.testing.tmpDir(.{});
    defer tmp_dir.cleanup();

    const config_content = "{}";
    try tmp_dir.dir.writeFile(.{ .sub_path = "test-config.json", .data = config_content });

    // Get real path
    const tmp_path = try tmp_dir.dir.realpathAlloc(allocator, ".");
    defer allocator.free(tmp_path);
    const explicit_path = try std.fs.path.join(allocator, &.{ tmp_path, "test-config.json" });
    defer allocator.free(explicit_path);

    // Test discovery
    const discovered = try discoverConfigPath(allocator, explicit_path);
    try std.testing.expect(discovered != null);
    defer allocator.free(discovered.?);

    try std.testing.expectEqualStrings(explicit_path, discovered.?);
}

test "discoverConfigPath returns null when no config exists" {
    const allocator = std.testing.allocator;

    // Create empty temp dir and change to it
    var tmp_dir = std.testing.tmpDir(.{});
    defer tmp_dir.cleanup();

    // Create .git to act as boundary
    try tmp_dir.dir.makeDir(".git");

    const original_cwd = try std.process.getCwdAlloc(allocator);
    defer allocator.free(original_cwd);

    const tmp_path = try tmp_dir.dir.realpathAlloc(allocator, ".");
    defer allocator.free(tmp_path);

    try std.posix.chdir(tmp_path);
    defer std.posix.chdir(original_cwd) catch {};

    // Test discovery (should return null)
    const discovered = try discoverConfigPath(allocator, null);
    try std.testing.expect(discovered == null);
}

test "upward search stops at .git boundary" {
    const allocator = std.testing.allocator;

    // Create directory structure:
    // temp/
    //   config.json (this should NOT be found)
    //   project/
    //     .git/
    //     subdir/ (we'll run from here)

    var tmp_dir = std.testing.tmpDir(.{});
    defer tmp_dir.cleanup();

    // Create config at root
    try tmp_dir.dir.writeFile(.{ .sub_path = ".complexityguard.json", .data = "{}" });

    // Create project with .git boundary
    try tmp_dir.dir.makeDir("project");
    try tmp_dir.dir.makeDir("project/.git");
    try tmp_dir.dir.makeDir("project/subdir");

    // Change to subdir
    const original_cwd = try std.process.getCwdAlloc(allocator);
    defer allocator.free(original_cwd);

    const tmp_path = try tmp_dir.dir.realpathAlloc(allocator, "project/subdir");
    defer allocator.free(tmp_path);

    try std.posix.chdir(tmp_path);
    defer std.posix.chdir(original_cwd) catch {};

    // Search should return null because .git boundary prevents finding parent config
    const discovered = try discoverConfigPath(allocator, null);
    try std.testing.expect(discovered == null);
}

File: src/cli/discovery.rs
use std::path::Path;

use super::config::Config;

/// Config file names to search for, in priority order.
///
/// For v0.8, only JSON files are parsed. TOML filenames are listed for
/// discovery order parity with the Zig binary but are skipped if found.
const CONFIG_FILENAMES: &[&str] = &[
    ".complexityguard.json",
    "complexityguard.config.json",
    ".complexityguard.toml",
    "complexityguard.config.toml",
];

/// Discover and load a config file.
///
/// If `explicit_path` is Some, load that file directly.
/// Otherwise, search upward from CWD through all parent directories,
/// stopping at a `.git` boundary or filesystem root.
/// Returns None if no config file is found.
pub fn discover_config(explicit_path: Option<&str>) -> anyhow::Result<Option<Config>> {
    if let Some(path) = explicit_path {
        let config = load_config_file(path)?;
        return Ok(Some(config));
    }

    // Upward search from CWD
    let cwd = std::env::current_dir()?;
    let mut search_dir = cwd.as_path();

    loop {
        // Check for .git boundary - stop searching above this directory
        // (we still check this directory itself before stopping)
        for filename in CONFIG_FILENAMES {
            let candidate = search_dir.join(filename);
            if candidate.exists() {
                // Skip TOML files in v0.8 (JSON only)
                if filename.ends_with(".toml") {
                    // Log would go here in verbose mode; continue searching for JSON
                    continue;
                }
                let config = load_config_file(candidate.to_string_lossy().as_ref())?;
                return Ok(Some(config));
            }
        }

        // Check for .git boundary - stop after checking this directory
        if search_dir.join(".git").exists() {
            break;
        }

        // Move to parent directory
        match search_dir.parent() {
            Some(parent) => search_dir = parent,
            None => break,
        }
    }

    Ok(None)
}

/// Load and parse a JSON config file from a specific path.
fn load_config_file(path: &str) -> anyhow::Result<Config> {
    let content = std::fs::read_to_string(path)
        .map_err(|e| anyhow::anyhow!("Failed to read config file '{}': {}", path, e))?;
    let config: Config = serde_json::from_str(&content)
        .map_err(|e| anyhow::anyhow!("Failed to parse config file '{}': {}", path, e))?;
    Ok(config)
}

/// Check whether a path looks like a config file we handle (JSON only in v0.8).
pub fn is_json_config(path: &Path) -> bool {
    matches!(
        path.file_name().and_then(|n| n.to_str()),
        Some(".complexityguard.json") | Some("complexityguard.config.json")
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;

    fn write_temp_config(dir: &tempfile::TempDir, filename: &str, content: &str) -> String {
        let path = dir.path().join(filename);
        let mut file = std::fs::File::create(&path).unwrap();
        file.write_all(content.as_bytes()).unwrap();
        path.to_string_lossy().to_string()
    }

    #[test]
    fn test_explicit_path_loads_config() {
        let dir = tempfile::tempdir().unwrap();
        let json = r#"{"output": {"format": "json"}}"#;
        let path = write_temp_config(&dir, "my-config.json", json);

        let result = discover_config(Some(&path)).unwrap();
        assert!(result.is_some());
        let config = result.unwrap();
        assert_eq!(config.output.unwrap().format, Some("json".to_string()));
    }

    #[test]
    fn test_explicit_path_missing_returns_error() {
        let result = discover_config(Some("/nonexistent/path/config.json"));
        assert!(result.is_err());
    }

    #[test]
    fn test_no_config_file_returns_none() {
        // Use a temp directory with no config files present
        // We can't easily test the upward search without changing CWD,
        // but we can test that discover_config handles missing files gracefully
        // by verifying is_json_config behavior
        let path = Path::new(".complexityguard.json");
        assert!(is_json_config(path));

        let path = Path::new("complexityguard.config.json");
        assert!(is_json_config(path));

        let path = Path::new(".complexityguard.toml");
        assert!(!is_json_config(path));

        let path = Path::new("other.json");
        assert!(!is_json_config(path));
    }

    #[test]
    fn test_load_config_file_valid_json() {
        let dir = tempfile::tempdir().unwrap();
        let json = r#"{"analysis": {"threads": 4}}"#;
        let path = write_temp_config(&dir, "config.json", json);

        let config = load_config_file(&path).unwrap();
        assert_eq!(config.analysis.unwrap().threads, Some(4));
    }

    #[test]
    fn test_load_config_file_invalid_json_returns_error() {
        let dir = tempfile::tempdir().unwrap();
        let path = write_temp_config(&dir, "config.json", "{ invalid json }");

        let result = load_config_file(&path);
        assert!(result.is_err());
        assert!(result.unwrap_err().to_string().contains("Failed to parse"));
    }

    #[test]
    fn test_load_config_file_all_sections() {
        let dir = tempfile::tempdir().unwrap();
        let json = r#"{
            "output": {"format": "sarif", "file": "out.sarif"},
            "analysis": {"threads": 8, "no_duplication": true},
            "files": {"include": ["src/**"], "exclude": ["**/*.test.ts"]},
            "weights": {"cognitive": 0.4},
            "baseline": 80.0
        }"#;
        let path = write_temp_config(&dir, "config.json", json);

        let config = load_config_file(&path).unwrap();
        assert_eq!(config.output.as_ref().unwrap().format, Some("sarif".to_string()));
        assert_eq!(config.output.as_ref().unwrap().file, Some("out.sarif".to_string()));
        assert_eq!(config.analysis.as_ref().unwrap().threads, Some(8));
        assert_eq!(config.analysis.as_ref().unwrap().no_duplication, Some(true));
        assert_eq!(
            config.files.as_ref().unwrap().include,
            Some(vec!["src/**".to_string()])
        );
        assert_eq!(config.baseline, Some(80.0));
    }
}
